### ISSUE TITLE
jump to function and autocomplete broken

### ISSUE THREAD

**Issue by LogSpider** [bash-lsp/bash-language-server#623]
If i install latest version (1.21.0) i can't jump to functions anymore. Also variables or other functions no longer known for autocompletion. If i use older Version (1.14.0) everything is working fine. Used with Remote SSH, Bash IDE, Git Graph under Linux.

![image](https://user-images.githubusercontent.com/96056194/207259335-a1de0d8c-2500-4006-82aa-c3c9c44325b0.png)

**Comment by LogSpider** [comment]
we have nested descending sourcing with thousands of functions, that is working perfectly with bash....that was working over years now with vscode bash ide addon. the last uptes broke it completle....if i make a simple script and source (include) a second, functions and vars are not known.

Example Main Script:
`# DEFINE FUNCTION PATH
funcPath="/my/function/path"

# LOAD FUNCTIONS
source $funcPath/global/its.func.script.sh
source $funcPath/its.func.$sName.sh
source $funcPath/its.func.somesubfuncs.sh`

Example Global Function Script (Loaded above in Main Script):
# SET BASIC VARS
export ROOTPATH="/my/root/path"
export SCRIPTPATH="$ROOTPATH/script"
export SCRIPTNAME="`basename "$0"`"
export SCRIPTARGUMENTS="$@"
export BINPATH="`dirname "$0"`"
export sName=$(echo $SCRIPTNAME | cut -f2 -d.)
cd $BINPATH
source $SCRIPTPATH/functions/global/its.func.logging.sh
source $SCRIPTPATH/functions/global/its.func.cout.sh
source $SCRIPTPATH/functions/global/its.func.cmd.sh
source $SCRIPTPATH/functions/global/its.func.customer.sh
source $SCRIPTPATH/functions/global/its.func.proc.sh
source $SCRIPTPATH/functions/global/its.func.net.sh
source $SCRIPTPATH/functions/global/its.func.check.sh
source $SCRIPTPATH/functions/global/its.func.mail.sh
source $SCRIPTPATH/functions/global/its.func.butler.sh
source $SCRIPTPATH/functions/global/its.func.sec.sh
source $SCRIPTPATH/functions/global/its.func.table.sh
source $SCRIPTPATH/properties/os/locale.conf
source $SCRIPTPATH/properties/os/path.conf
source $SCRIPTPATH/properties/os/color.conf

**Comment by LogSpider** [comment]
@skovhus it is not working anymore since version 1.16.2, jumped versions up from 1.14.0 to 1.16.2....since version 1.16.2 strg + mouse1 or anything else like autocompletion dont work anymore.

**Comment by skovhus** [comment]
Thanks for the detailed response! 

Can you try turning on the `bashIde.includeAllWorkspaceSymbols` configuration flag? That configuration flag is meant for supporting the old behaviour of resolving all workspace symbols.

<img width="549" alt="Screenshot 2022-12-13 at 11 48 20" src="https://user-images.githubusercontent.com/1260305/207297880-2c5cb0ea-3a56-4985-87a2-efc513b950e7.png">

Resolving all workspace symbols as a default introduced a lot of false positives for people. So in the latest release we now default to parse (non dynamic) source statements. But I do see that the new behaviour can cause issues... It is a bit of a tradeoff. We might be able to parse dynamic source statements (i.e. using variables as you have here). Or revert to the old behaviour for files that use dynamic source statements. 🤔

**Comment by LogSpider** [comment]
Okay, i understand now. With Dynamic you mean sourcing / including with Vars? But why should this be a problem? When i use bash in console and make . /mainscript.sh, all vars and functions are available, even if i use relative paths or variables in source / include declarations. That would mean, when i set absolut pathes without vars, when including other scripts, it should work.

**Comment by skovhus** [comment]
> @skovhus it is not working since version 1.16.2, jumped versions up from 1.14.0 to 1.16.2....since version 1.16.2 ctrl+ mouse1 for jumping to function or anything else like autocompletion of vars dont work anymore.

I see now that you are probably also affected by a change we did to minimize the CPU usage analyzing unnessary bash files. In 1.16.2 the background analysis was limited to 500 bash files. But can be tweaked using the `bashIde.backgroundAnalysisMaxFiles` configuration option. I assume this is why you see missing completion and jump to definition even before the new sourcing aware symbol completion was introduced.

**Comment by LogSpider** [comment]
i already changed that to 9999 without any change xD

**Comment by skovhus** [comment]
> Dynamic you mean sourcing / including with Vars? 

Dynamic sourcing meaning including variables, yes. 

> But why should this be a problem? When i use bash in console and make . /mainscript.sh, all vars and functions are available, even if i use relative paths or variables in source / include declarations. 

It is simply not implemented. The bash language server uses bash files ASTs, documentation available from the system, etc, but currently doesn't resolve dynamic sourcing statements as this means we would need to either execute part of the bash program or implement our own interpreter. 

> That would mean, when i set absolut pathes without vars, when including other scripts, it should work.

Yes using absolute paths should make completion and jump to definition work in the latest version without enabling the `includeAllWorkspaceSymbols` flag.

**Comment by LogSpider** [comment]
i enabled it on user, remote and workspace...still doesn't work. i know you will not believe it :/

![image](https://user-images.githubusercontent.com/96056194/207304108-5985eef6-e492-4128-88df-61f70ca49b4b.png)

**Comment by skovhus** [comment]
If it still doesn't work after turning on `includeAllWorkspaceSymbols` then I would love to see the log lines from the bash language server. You can find those under `View > Output > Bash IDE`. Specifially the lines around the background analysis. 

```
BackgroundAnalysis: resolving glob "**/*@(.sh|.inc|.bash|.command)" inside "file:///Users/kenneth/git/bash-language-server"...
BackgroundAnalysis: Glob resolved...
```

**Comment by skovhus** [comment]
I might have found an issue here. The background analysis received a workspace folder with the `file://` prefix that it silently fail on and doesn’t resolve any files.

**Comment by LogSpider** [comment]
I think you are able to fix it, i understand what happened. Thank you very much for fast replies.

**Comment by skovhus** [comment]
A fix is out, bash-language-server@4.1.1 and the vscode extension 1.22.0. Does it work for you?

**Comment by LogSpider** [comment]
Yes, it is working with includeAllWorkspaceSymbols.

```
BackgroundAnalysis: Glob resolved with 84 files after 0.131 seconds
BackgroundAnalysis: Completed after 0.705 seconds.
[Warn  - 10:46:59] getExplainshellDocumentation exception: Only absolute URLs are supported
```

Thx again.

**Comment by skovhus** [comment]
Thanks!
